[RF-DOCS] Rewrite and restructure the Action Cable Overview guide (#58262)
rails/rails: guides/source/action_cable_overview.md

Action Cable Overview
=====================

In this guide, you will learn how to use Action Cable to add real-time bi-directional communication to your Rails application.

After reading this guide, you will know:

* What Action Cable is and how it works.
* Action Cable core concepts such as Connections, Channels, and Streams.
* How to use the JavaScript client library.
* The pub/sub adapters available for Action Cable.
* How to configure and deploy Action Cable.

--------------------------------------------------------------------------------

What is Action Cable?
---------------------

Action Cable provides a mechanism for real-time bi-directional communication with the web browser by pairing [WebSockets](https://en.wikipedia.org/wiki/WebSocket) with [Pub/Sub](https://en.wikipedia.org/wiki/Publish-subscribe_pattern). Using it, we can receive and broadcast data to the browser in real-time with ease.

### WebSockets

> The WebSocket protocol enables bi-directional communication between a browser and server over a persistent TCP connection. Either side can send data at any time — no explicit request for data is required. Common use cases include chat, live notifications, live dashboards, or collaborative editing. It's standardized in [RFC 6455](https://datatracker.ietf.org/doc/html/rfc6455)

Reviewer: Missing terminal period.

```suggestion
The WebSocket protocol enables bi-directional communication between a browser and server over a persistent TCP connection. Either side can send data at any time — no explicit request for data is required. Common use cases include chat, live notifications, live dashboards, or collaborative editing. It's standardized in [RFC 6455](https://datatracker.ietf.org/doc/html/rfc6455).
```

A WebSocket connection starts with an HTTP `GET` request asking the server to upgrade the connection to a WebSocket. The server will then repond with a `101 Switching Protocols` status code and the connection will switch over to a WebSocket. This is known as a _handshake_.

![Sequence diagram depicting the WebSocket handshake](images/action_cable/handshake.png)

WebSocket connections in the browser are managed using a [JavaScript API](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket). Action Cable provides a JavaScript library to abstract some of the complexity.

### Pub/Sub

[Pub/Sub](https://en.wikipedia.org/wiki/Publish-subscribe_pattern) is shorthand for _publish-subscribe_. It's a software messaging paradigm where entities can _subscribe_ to data streams. Other entities, called _publishers_, can then publish data to these streams for delivery to all subscribers.

This paradigm makes it easy to broadcast data based on _topics_, rather than sending data to specific receivers.

> Action Cable doesn't implement pub/sub itself, instead it relies on an external service via an adapter. The default is [Solid Cable](https://github.com/rails/solid_cable) which uses SQLite for message storage and delivery. Adapters for PostgreSQL and Redis are also available.

Reviewer: This conflicts with L809 ("uses Active Record under the hood and has been tested with MySQL, SQLite, and PostgreSQL"). SQLite is the new-app default, not what Solid Cable itself uses.

```suggestion
Action Cable doesn't implement pub/sub itself, instead it relies on an external service via an adapter. The default is [Solid Cable](https://github.com/rails/solid_cable), which stores and delivers messages through your database via Active Record. Adapters for PostgreSQL and Redis are also available.
```

```ruby
# app/channels/application_cable/connection.rb

module ApplicationCable
  class Connection < ActionCable::Connection::Base
    identified_by :current_user

    def connect
      self.current_user = authenticate_user
    end

    private
      def authenticate_user
        # Authenticate the user using cookies.

        # Reject the connection using `reject_unauthorized_connection`
        # if authentication fails.
      end
  end
end
```

An instance of `ApplicationCable::Connection` is created for each connection. Creating a consumer on the client will trigger the `connect` method shown above. Use cookies to authenticate the user, rejecting the connection if it fails. `current_user` is set on the connection to uniquely identify it.

The Rails `session` cannot be accessed in this context since `ApplicationCable::Connection` doesn't inherit from `ActionController::Base`, and hence the modules related to the session are not included. If your session data is stored in a cookie, you can access it using:

```ruby

> cookie_name = Rails.app.config.session_options[:key]

Reviewer: `Rails.app.config` works (`alias :app :application`, [rails.rb:47](https://github.com/rails/rails/blob/main/railties/lib/rails.rb#L47)), but it's unusual for the guides — `configuring.md` uses `Rails.application.config` 14×, and `Rails.app.config` appears nowhere else in the repo. Also at L939, L972 and L1000.

```suggestion
cookie_name = Rails.application.config.session_options[:key]
```

Author: The `Rails.app` alias is new on Rails `main` I believe, hence why none of the guides use it yet. It's more concise so I'd like to start using it now as the guides need to be relevant as per `main`. 

The configuring guide hasn't yet been updated as part of this project, so when we update that guide we will change to `Rails.app`.

const consumer = createConsumer()

consumer.subscriptions.create(
  {
    channel: "ChatChannel",
    room: "general"
  },
  {
    connected() {
      console.log("Subscribed to ChatChannel")
    },

    disconnected() {
      console.log("Unsubscribed from ChatChannel")
    },

    received(data) {
      console.log(data.body)
    }
  }
)
```

Since WebSockets are bi-directional, clients can send data to the server over the same connection.

```js#11
import { createConsumer } from "@rails/actioncable"
const consumer = createConsumer()

const generalChat = consumer.subscriptions.create(
  {
    channel: "ChatChannel",
    room: "general"
  },
  {
    connected() {
      generalChat.send({ body: "Hi, I'm online!" })
    }
  }
)
```

This message is received in the `ChatChannel`.

```ruby
# app/channels/chat_channel.rb

class ChatChannel < ApplicationCable::Channel
  # ...

  def receive(data)
    # Handle received message
  end
end
```

A common use case is to rebroadcast data received from the client back to all the subscribers.

```ruby
# app/channels/chat_channel.rb

class ChatChannel < ApplicationCable::Channel
  # ...

  def receive(data)

>     ActionCable.server.broadcast("chat_#{params[:room]}", data["body"])

Reviewer: This broadcasts a bare string, which doesn't match the client that receives it.

```suggestion
    ActionCable.server.broadcast("chat_#{params[:room]}", data)
```

The subscriber at L178–180 does `received(data) { console.log(data.body) }`. The client sends `{ body: "Hi, I'm online!" }` (L198), so `receive` gets `{"body" => "..."}` — broadcasting `data["body"]` unwraps it to a plain string and `data.body` is `undefined` on the receiving end. The first broadcast example (L155) keeps the hash shape, so passing `data` through also makes the two consistent.

The "this will also broadcast back to the initial sender" note below still applies unchanged.

  end
end
```

Note that this will also broadcast the message back to the initial sender. Depending on your use case, you may wish to add client-side logic to ignore the duplicate message.

Clients can also trigger actions in the channel.

```ruby
# app/channels/chat_channel.rb

class ChatChannel < ApplicationCable::Channel
  # ...

  def appear
    current_user.appear
  end
end
```

```js
import { createConsumer } from "@rails/actioncable"
const consumer = createConsumer()

const generalChat = consumer.subscriptions.create(
  {
    channel: "ChatChannel",
    room: "general"
  },
  {
    connected() {
      // Triggers the `appear` method in `ChatChannel`
      generalChat.perform("appear")
    }
  }
)
```

That covers the basics of creating a WebSocket connection and sending and receiving messages using Action Cable. In the next sections, we'll look more closely at the server components and client library.

Server Components
-----------------

Rails doesn't create Action Cable files when a new project is generated. Create them by generating a new channel:

```bash

> $ bin/rails generation channel chat

Reviewer: `generation` → `generate`.

```suggestion
$ bin/rails generate channel chat
```

Worth catching because it's the first command a reader runs, and it's load-bearing for this section — the paragraph above establishes that `app/channels` isn't created for you, so everything after depends on this line working.

* `app/channels/chat_channel.rb` <br>
  The channel class itself.

A test file and some JavaScript files will also be created. See the next section on the [JavaScript Client Library](#javascript-client-library) to learn more about the generated JavaScript files and the [Testing guide](testing.html#testing-action-cable) for guidance on writing tests for Action Cable channels.

### Connection

This is the default `Connection` file generated by Rails:

```ruby
# app/channels/application_cable/connection.rb

module ApplicationCable
  class Connection < ActionCable::Connection::Base
  end
end
```

An instance of this class represents every WebSocket connection. As demonstrated in the previous section, authenticate the connection before accepting it:

```ruby
# app/channels/application_cable/connection.rb

module ApplicationCable
  class Connection < ActionCable::Connection::Base
    identified_by :current_user

    def connect
      # Authenticate the connection, using `reject_unauthorized_connection` to
      # terminate it if authentication fails.
    end

    def disconnect
      # Any cleanup work needed when the cable connection is cut.
    end
  end
end
```

> [`identified_by`](https://api.rubyonrails.org/classes/ActionCable/Connection/Identification/ClassMethods.html#method-i-identified_by) accepts multiple arguments if you would like to make multiple objects available to the connection. It serves as the Action Cable equivalent of [`ActiveSupport::CurrentAttributes`](https://api.rubyonrails.org/classes/ActiveSupport/CurrentAttributes.html). You must set at least one value declared by `identified_by` for the connection to succeed.

Reviewer: This constraint isn't enforced by the framework. `handle_open` ([`connection/base.rb:91`](https://github.com/rails/rails/blob/main/actioncable/lib/action_cable/connection/base.rb#L91)) only rescues `UnauthorizedError` from `reject_unauthorized_connection`, and `subscribe_to_internal_channel` no-ops when `connection_identifier` is blank. The default generated `connection.rb` declares no identifiers at all and connects fine.

Suggest describing what's lost instead:

```suggestion
[`identified_by`](https://api.rubyonrails.org/classes/ActionCable/Connection/Identification/ClassMethods.html#method-i-identified_by) accepts multiple arguments if you would like to make multiple objects available to the connection. It serves as the Action Cable equivalent of [`ActiveSupport::CurrentAttributes`](https://api.rubyonrails.org/classes/ActiveSupport/CurrentAttributes.html). A connection with no identifiers set is still accepted, but it won't have a `connection_identifier`, so it can't be located or disconnected via `ActionCable.server.remote_connections`.
```

Cookies are not available in a WebSocket connection so any user-specific data for authorization or other uses must be stored in `identified_by` attributes.

Complete usage information is available in the [API docs](https://api.rubyonrails.org/classes/ActionCable/Connection/Base.html).

#### Connection Callbacks

Similar to controllers, [callbacks](https://edgeapi.rubyonrails.org/classes/ActionCable/Connection/Callbacks/ClassMethods.html) are available in the `Connection` which are invoked when sending commands to the client, such as subscribing, unsubscribing, or performing an action:

```ruby#7-9
# app/channels/application_cable/connection.rb

module ApplicationCable
  class Connection < ActionCable::Connection::Base
    identified_by :current_user

    before_command  ->(conn) { logger.debug "Before command for #{conn.current_user.id}" }
    around_command  :set_current_account
    after_command   ->(conn) { logger.debug "After command for #{conn.current_user.id}" }

    # ...

    private
      def set_current_account

>         Current.set(account: user.account) { yield }

Reviewer: `user` is undefined here, so this raises `NoMethodError` as written.

```suggestion
        Current.set(account: current_user.account) { yield }
```

This is adapted from the docs on [`connection/callbacks.rb`](https://github.com/rails/rails/blob/main/actioncable/lib/action_cable/connection/callbacks.rb), where the connection declares `identified_by :user`. The snippet here declares `identified_by :current_user` (L339), so the identifier was renamed but the callback body wasn't.

When a consumer triggers a subscription to a channel, a new channel instance is created and the `subscribed` method is called. Here, you can do any initial setup as well as wire up any streams you would like to send to the client.

```ruby
# app/channels/chat_channel.rb

class ChatChannel < ApplicationCable::Channel
  def subscribed
    @room = Chat::Room[params[:room_id]]
    return reject unless current_user.can_access?(@room)

    current_user.appear
    stream_for @room
  end

  # ...
end
```

Ensure you do an authorization check so users only access permitted resources. Use `reject` to disallow the subscription.

Set up a stream for subscribers using [`stream_from`][] or [`stream_for`][]. [`stream_from`][] only accepts a string as the stream name:

```ruby
# app/channels/chat_channel.rb

class ChatChannel < ApplicationCable::Channel
  def subscribed
    # ...

    room_name = params[:room]
    stream_from "room_#{room_name}"
  end

  # ...
end
```

[`stream_for`][] accepts one or more _broadcastables_, which are objects that respond to `to_gid_param` or `to_param`. Active Records objects respond to both these methods. `stream_for` will serialize them into a string.

> ```ruby#7

Reviewer: The highlight lands on `@room = Chat::Room[...]`, but the paragraph is about `stream_for`, which is line 8 of the block.

````suggestion
```ruby#7-8
````

  # sent by the client.
  def speak(data)
    room_id = params[:room_id]
    logger.debug "Data received for #{room_id}: #{data}"

    @room.speak(data, user: current_user)
  end
end
```

```js
import { createConsumer } from "@rails/actioncable"
const consumer = createConsumer()

const generalChat = consumer.subscriptions.create(
  {
    channel: "ChatChannel",
    room_id: "667ea191042b8303e920"
  },
  {
    connected() {
      // Triggers the `speak` method in `ChatChannel`
      generalChat.perform("speak", { message: "Hello!" })
    }
  }
)
```

#### Handling Exceptions

As with `ApplicationCable::Connection`, use [`rescue_from`][] in a channel to handle exceptions:

```ruby
# app/channels/chat_channel.rb
class ChatChannel < ApplicationCable::Channel
  rescue_from "MyError", with: :deliver_error_message

  private
    def deliver_error_message(error)
      # Handle error ...
    end
end
```

#### Channel Callbacks

A channel also provides [callbacks](https://api.rubyonrails.org/classes/ActionCable/Channel/Callbacks.html), similar to controllers and `Connection`, that are invoked during the lifecycle of a channel:

```ruby

> # app/channels.chat_channel.rb

Reviewer: Dot where a slash belongs in the path comment.

```suggestion
# app/channels/chat_channel.rb
```

class ChatChannel < ApplicationCable::Channel
  before_subscribe :authorize_subscription
  # `after_subscribe` is aliased as `on_subscribe`
  after_subscribe :appear, unless: :subscription_rejected?

  before_unsubscribe :disappear

>   # `after_unsubscribed` is aliased as `on_unsubscribe`

Reviewer: Method name typo in the comment — the callback on the line below is correct.

```suggestion
  # `after_unsubscribe` is aliased as `on_unsubscribe`
```

Per [`channel/callbacks.rb`](https://github.com/rails/rails/blob/main/actioncable/lib/action_cable/channel/callbacks.rb) the alias pairs are `after_subscribe`/`on_subscribe` and `after_unsubscribe`/`on_unsubscribe`.

createConsumer(getWebSocketURL)

function getWebSocketURL() {
  const token = localStorage.getItem('auth-token')
  return `wss://example.com/cable?token=${token}`
}
```

NOTE: `wss` means the WebSocket connection is secured with encryption, similar to how `https` denotes an encrypted `http` connection.

You can globally set the url for the consumer by adding a `<meta>` tag named `action-cable-url` to your document's `<head>`. The [`action_cable_meta_tag`] helper method can be used for this.

See the [mount paths](#mount-path-and-url) section for details on mounting Action Cable on a different path, and the [deployment](#deployment) section to learn how to setup standalone Action Cable servers.

### Channels

The default JavaScript channel generated by Rails is:

```js
// app/javascript/channels/chat_channel.js

import consumer from "channels/consumer"

consumer.subscriptions.create("ChatChannel", {
  connected() {
    // Called when the subscription is ready for use on the server
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
  },

  received(data) {
    // Called when there's incoming data on the websocket for this channel
  }
});

> ````

Reviewer: Stray fourth backtick closing the `js` fence. I ran this through the guides' own Redcarpet config ([`markdown.rb:72`](https://github.com/rails/rails/blob/main/guides/rails_guides/markdown.rb#L72)) and it renders correctly, so this is cosmetic only — flagging for consistency.

````suggestion
```
````

```js
import consumer from "channels/consumer"

consumer.subscriptions.create({ channel: "ChatChannel", room: "general" })
consumer.subscriptions.create({ channel: "ChatChannel", room: "watercooler" })
```

Each subscription will create a new channel instance on the server.

#### Sending Messages and Triggering Remote Actions

After a consumer has subscribed to a channel, it can send messages to the server:

```js
import consumer from "channels/consumer"

const generalRoom = consumer.subscriptions.create(
  {
    channel: "ChatChannel",
    room: "general"
  },
  {
    connected() {
      // This will trigger the `receive` method in the
      // `ChatChannel` on the server.
      generalRoom.send({ message: "I'm here!" })
    }
  }
)
```

And also trigger remote actions in that channel:

```js
import consumer from "channels/consumer"

const generalRoom = consumer.subscriptions.create(
  {
    channel: "ChatChannel",
    room: "general"
  },
  {
    connected() {
      // Triggers the `speak` method in `ChatChannel`
      generalRoom.perform("speak", { message: "Hello!" })
    }
  }
)
```

Adapters
--------

> Action Cable relies on external service for _pub/sub_. Rails interacts with the service via an adapter and this must be configured in `config/cable.yml`.

Reviewer: ```suggestion
Action Cable relies on an external service for _pub/sub_. Rails interacts with the service via an adapter and this must be configured in `config/cable.yml`.
```

### Redis

Action Cable's Redis adapter integrates with [Redis' pub/sub messaging system](https://redis.io/docs/latest/develop/pubsub/).

NOTE: Redis pub/sub is independent of its key-value database. They operate in the same process and share networking, but pub/sub operations do not affect the key-value database in any way.

Configure the Redis adapter by defining the URL to the Redis server as shown below:

```yaml
# config/cable.yml

production:
  adapter: redis
  url: redis://redis.infra.local:6379/1
  channel_prefix: my_app_production
```

You can optionally define a `channel_prefix` to avoid channel name collisions when using the same Redis server for multiple applications.

The Redis adapter supports SSL/TLS connections:

```yaml#5,7,8
# config/cable.yml

production:
  adapter: redis
  url: rediss://redis.infra.local:6379/1
  channel_prefix: my_app_production
  ssl_params:
    ca_file: "/path/to/ca.crt"
```

The `ssl_params` are passed directly to [`OpenSSL::SSL::SSLContext#set_params`](https://docs.ruby-lang.org/en/master/OpenSSL/SSL/SSLContext.html#method-i-set_params) and can be any valid attribute of the SSL context.

### PostgreSQL

> The PostgreSQL adapter uses the [`NOTIFY`](https://www.postgresql.org/docs/current/sql-notify.html) and [`LISTEN`](https://www.postgresql.org/docs/18/sql-listen.html) commands to implement pub/sub using the PostgreSQL databse. It uses Active Record's connection pool, and thus the application's `config/database.yml` database configuration.

Reviewer: `databse` typo, and the two PostgreSQL links point at different doc versions (`current` vs `18`).

```suggestion
The PostgreSQL adapter uses the [`NOTIFY`](https://www.postgresql.org/docs/current/sql-notify.html) and [`LISTEN`](https://www.postgresql.org/docs/current/sql-listen.html) commands to implement pub/sub using the PostgreSQL database. It uses Active Record's connection pool, and thus the application's `config/database.yml` database configuration.
```

```ruby
# config/environments/production.rb

config.action_cable.allowed_request_origins = ["https://my-app.com", %r{https://*.my-app.com}]
```

Allow requests from any origin by disabling request forgery protection:

```ruby
# config/environments/production.rb

config.action_cable.disable_request_forgery_protection = true
```

### Mount Path and URL

Action Cable is _mounted_ at `/cable` by default. This means the JavaScript `consumer` makes a request to `/cable` to initiate the WebSocket handshake. You can change the mount path in your configuration:

```ruby
# config/initializers/action_cable.rb

Rails.app.config.action_cable.mount_path = "/websocket"
```

When running a [standalone Action Cable server](#standalone-action-cable-server), configure the URL to the server using:

```ruby
# config/environments/production.rb

config.action_cable.mount_path = nil
config.action_cable.url = "wss://cable.my-rails-app.com"
```

> Set the Action Cable URL for the JavaSript consumer by adding [`action_cable_meta_tag`][] to your application's `<head>`:

Reviewer: ```suggestion
Set the Action Cable URL for the JavaScript consumer by adding [`action_cable_meta_tag`][] to your application's `<head>`:
```

```erb
<%# app/views/layouts/application.html.erb %>

<head>
  <%= action_cable_meta_tag %>
</head>
```

This will write the Action Cable URL in a `<meta>` tag which will be automatically read by Action Cable's JavaScript client to create a consumer pointing to the custom URL.

[`action_cable_meta_tag`]: https://api.rubyonrails.org/classes/ActionCable/Helpers/ActionCableHelper.html#method-i-action_cable_meta_tag

### Worker Pool

Action Cable maintains its own thread pool which is distinct from your application server's thread pool. It's exclusively used to process WebSocket messages for Action Cable without affecting HTTP requests. The size of the thread pool can be configured using:

```ruby
# config/initializers/action_cable.rb

Rails.app.config.action_cable.worker_pool_size = 4
```

> The default value is `4`. Each worker requires it's own database connection. A worker pool size of `4` requires 4 database connections. Ensure the `pool` attribute in your `config/database.yml` is high enough to allow both Action Cable's threads and your application server's threads to hold database connections.

Reviewer: ```suggestion
The default value is `4`. Each worker requires its own database connection. A worker pool size of `4` requires 4 database connections. Ensure the `pool` attribute in your `config/database.yml` is high enough to allow both Action Cable's threads and your application server's threads to hold database connections.
```

```

For a full list of configuration options, see the [`ActionCable::Configuration`](https://api.rubyonrails.org/classes/ActionCable/Configuration.html) class.

Deployment
----------

Action Cable can be run within your main Rails server's process. It will be mounted at `/cable` by default and can accept and process WebSocket connections alongside HTTP requests. No additional configuration is required for this setup.

For apps handling a higher level of traffic, you may wish to run standalone Action Cable Servers.

### Standalone Action Cable Server

A standalone Action Cable server boots your Rails app as normal, but only processes Action Cable connections and no HTTP requests.

The below Rack configuration demonstrates how to start an Action Cable server:

```ruby
# cable.ru

require_relative "../config/environment"
Rails.application.eager_load!

run ActionCable.server
```

Start the server using:

```bash
$ bundle exec puma -p 28080 cable.ru
```

Your Action Cable server will be started on port 28080.

Ensure you [configure Rails to point to the Action Cable server's URL](#mount-path-and-url).

> ## Testing

Reviewer: Every other section header in this file uses Setext underlines; this is the only ATX one.

```suggestion
Testing
-------
```
